src/main/scala/GameHelpers.scala

```diff
@@ -0,0 +1,69 @@
+import scala.util.Random
+
+object GameHelpers {
+
+  def fillRack(rack: List[Tile], letterBag: List[Tile]): (List[Tile], List[Tile]) = {
+    val (drawnTiles, newBag) = letterBag.splitAt(7 - rack.length)
+    val filledRack = rack ++ drawnTiles
+
+    if (filledRack.length < 7) {
+      println("\u001b[0m---------")
+      print("\u001b[31;1mWARNING: ")
+      println("\u001b[36mThe letter bag is empty.")
+      print("\u001b[0m")
+    }
+
+    (filledRack, newBag)
+  }
+
+  def removeTilesFromRack(word: String, rack: List[Tile]): List[Tile] = {
+    val wordTiles: List[Tile] = word.toList.map(c => Tile(c))
+    rack diff wordTiles
+  }
+
+  def calculateWordScore(word: String): Int = {
+    word.toList.map(Tile(_).value).sum
+  }
+
+  def isInDictionary(word: String): Boolean = {
+    if (Dictionary.all.contains(word.toLowerCase())) true else false
+  }
+
+  def isInRack(word: String, rack: List[Tile]): Boolean = {
+    val sortedWord = word.toList.sortWith(_.compareTo(_) < 0)
+    val sortedRack = rack.map(_.letter).sortWith(_.compareTo(_) < 0)
+
+    def loop(wL: List[Char], wI: Int, rL: List[Char], rI: Int): Boolean = {
+      if (wI == sortedWord.length) true
+      else if (rI == sortedRack.length) false
+      else if (wL(wI) == rL(rI)) loop(wL, wI + 1, rL, rI + 1)
+      else if (wL(wI) != rL(rI)) loop(wL, wI, rL, rI + 1)
+      else false
+    }
+
+    loop(sortedWord, 0, sortedRack, 0)
+  }
+
+  def isValidWord(word: String, rack: List[Tile]): Boolean = {
+    if(!isInDictionary(word)) { println(s"'$word' is not in the Scrabble Dictionary.")}
+    if(!isInRack(word, rack)) { println("You don't have those letter tiles in your rack.")}
+    isInDictionary(word) && isInRack(word, rack)
+  }
+
+  def playWord(rack: List[Tile], level: String): Option[String] = {
+    val possibleWords = Dictionary.all.filter(word => isInRack(word.toUpperCase, rack))
+    if (possibleWords == Nil) {
+      None
+    } else {
+      level match {
```

Comment:
you could do something along these lines to split out the common parts which keeping the structure a bit flatter
```scala
Some(possibleWords).filter(_.nonEmpty).map { possibleWords => 
  level match {
    case "easy" => Random.shuffle(possibleWords)
    case "hard" => possibleWords.sortBy(word => (word.length)).reverse
  }
}.map(_.head.toUpperCase)
```

Reply:
🤔  not sure I understand what `Some(possibleWords).filter(_.nonEmpty).map` is doing

Reply:
so to answer your specific question it's returning `None` if the possible words is empty otherwise it's returning a `Some` with whatever the output of the function inside the map is.

In real terms I just noticed your function is returning an Option, so rather than starting with if statements and spreading the Some/None around, it places the option up front and then processes the thing that's inside it (we don't have to keep mentioning Some or None after that point).

The only benefit is this is a way of restricting what the code can do to aid predictability and testability - as soon as you see "map" you know that the result must remain a "Some" without having to look inside or test the code inside.  The same goes for "filter" you know the only thing it can do is convert to None.

Having said all that, this is a typical functional programming way of doing abstraction and it can be less descriptive, so that's why I wrote you "could" do that rather than you "should"!  Grab me if you want any clarification!